### PR TITLE
Abstract flexbox over LayoutTree

### LINKED ISSUE #182: Introduce some tree abstraction
## What problem does this solve or what need does it fill?
If I want to use this library in my project I need to build and maintain a tree of layout nodes even when I have
some tree already.

## What solution would you like?
I propose introducing a trait, called `LayoutTree` with methods for traversing & accessing the styles. Something like this:
image

## What alternative(s) have you considered?
morphorm is already using similar abstraction, just in a little different flavor. https://github.com/vizia/morphorm/blob/main/src/hierarchy.rs

## Additional context
I think there is also another option, in a sense of visitor pattern, so instead of implementing some `Tree` trait, one could implement `TreeVisitor` which could look like this:

```
pub trait LayoutVisitor {
  fn open_node(&mut self, style: &LayoutStyle, dirty_flags: Xxx) -> bool;

  fn close_node(&mut self, result: &mut LayoutResult);

  fn visit_text(&mut self, paragraph: &dyn Paragraph, result: &mut LayoutResult);
}
```

This way we wouldn't need any `Tree::NodeRef` because the client is rather supposed to implement internal iteration for this library, including writing the results which means we don't need any id/ref and data can be stored anywhere/anyhow.

BTW: original discussion on reddit https://www.reddit.com/r/rust/comments/v9en4v/comment/ic6pfx9/?utm_source=share&utm_medium=web2x&context=3

### PR DESCRIPTION
**Pull request by jkelleyrtp**
This PR makes the flexbox algorithm generic over a LayoutTree trait.

At the time of filing this PR, the LayoutTree trait mostly abstracts over the broadest set of functions that the `Taffy` struct satisfies. I don't know or think this is the best and most general solution to flexboxing: Taffy internally does caching of calculations that a user's LayoutTree shouldn't really be concerned about. Perhaps some of the state - like caching - could remain in a Taffy struct, but this level of specificity intrinsically ties a user's UI system to Taffy due to dependent data structures.

Generally, since Taffy uses its own types, most implementations will be tied to Taffy anyways, so perhaps this linking of user code to Taffy is not a big deal.

Also - this trait supplies a method that returns a FlexboxLayout (not the final layout, despite the confusing name) which might not be exactly a trait implementation would return the "current state" of the DOM Node. For instance, in Dioxus, we are using the CSS Cascade to supply the "current state" of the DOM node's style. I think before publishing out these changes, it might be worth further abstracting the trait to supply single-level Taffy attributes. Perhaps a "TaffyAttribute" enum could be supplied so use cases like Dioxus' cascading CSS could be implemented efficiently.

Closes #182

### CODE REVIEW

File: src/tree.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    fn parent(&self, node: Node) -> Option<Node>;
+
+    // todo: allow abstractions over this so we don't prescribe how layout works
+    // for reference, CSS cascades require context, and storing a full flexbox layout for each node could be inefficient
+    //
+    /// Get the [`FlexboxLayout`] for this Node.
+    fn style(&self, node: Node) -> &FlexboxLayout;
+
+    /// Get the node's output "Final Layout"
+    fn layout(&self, node: Node) -> &Layout;
```
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply src/tree.rs:29]: Consider thoughts of this nature solicited by default! This is great feedback.

File: src/tree.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    fn child(&self, node: Node, index: usize) -> Node;
+
+    /// Get any available parent for this node
+    fn parent(&self, node: Node) -> Option<Node>;
+
+    // todo: allow abstractions over this so we don't prescribe how layout works
+    // for reference, CSS cascades require context, and storing a full flexbox layout for each node could be inefficient
+    //
+    /// Get the [`FlexboxLayout`] for this Node.
+    fn style(&self, node: Node) -> &FlexboxLayout;
```
Comment (nicoburns) [inline src/tree.rs:26]: I would like to suggest that `FlexboxLayout` is renamed to `FlexboxStyle`: the idea being to establish a naming convention that `*Style` is input and `*Layout` is output. `FlexboxLayout` is already in a file called `style.rs` and returned from a method called `style`, so this seems to improve naming consistency.

**Comment by nicoburns** [comment]
Is it worth thinking about how this will interact with supporting multiple layout algorithms? From some preliminary reading of the code, I believe we will probably want to split out the `Taffy` struct into something which simply holds data (i.e. something that implements `LayoutTree` here, and an entirely separate struct/module (possibly stateless) that implements the flexbox algorithm.

File: src/tree.rs (hunk 2 above)
Comment (Weibye) [reply src/tree.rs:26]: This makes a lot of sense, it is what it should have been all along.

Comment (alice-i-cecile) [reply src/tree.rs:26]: I'm on board with that. IMO we should do it in a small seperate PR though.

**Comment by geom3trik** [comment]
On the subject of supporting multiple layout algorithms, since I know that Morphorm is being looked at as a candidate, this isn't too far away from what Morphorm does with its `Hierarchy` trait. The main difference being that Morphorm requires the user to provide iterators over the whole tree while this approach simply queries nodes for their parent and children, which I think is much better. I'm going to see if I can adapt Morphorm to work in the same way. The tricky part for integrating multiple algorithms is going to be the mismatch between the style properties that each is using I think.

**Comment by alice-i-cecile** [comment]
> The tricky part for integrating multiple algorithms is going to be the mismatch between the style properties that each is using I think.

Agreed. We should start with an MVP where we can't mix styles in the same tree. Then once that's built and we have a couple candidates, we can start looking to Flutter to figure out if there's a good interface by which we can let them talk.

File: src/tree.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    /// Node needs to be measured
+    fn needs_measure(&self, node: Node) -> bool;
+
+    /// Get the primary cache for this Node.
+    ///
+    /// Taffy caches results of computations for nodes to not need re-caculating nodes it already knows
+    ///
+    /// When a node does not have a cache, Taffy will layout that node appropriately.
+    fn primary_cache(&mut self, node: Node) -> &mut Option<Cache>;
```
Comment (bestouff) [inline src/tree.rs:48]: I do not understand how this cache works (when is it invalidated, etc.).

File: src/tree.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    prelude::*,
+};
+
+/// Any item that implements the LayoutTree can be layed out using Taffy's algorithms.
+///
+/// Generally, Taffy expects your Node tree to be indexable by stable indicies. A "stable" index means that the Node's ID
+/// remains the same between re-layouts.
+pub trait LayoutTree {
+    /// Get the list of children IDs for the given node
+    fn children(&self, node: Node) -> &[Node];
```
Comment (bestouff) [inline src/tree.rs:14]: Maybe here you restrict too much the type of the underlying `children` set (which now *has* to be a `Vec` or something like that) and it's be better to have something like
```rust
     fn children(&self, node: Node) -> impl Iterator<Item=Node>;
```
(which also has a trivial implementation). Now if the `LayoutTree` implementor uses e.g. a `BTreeMap` as the underlying storage it still can trivially implement `children()`.

File: src/tree.rs (hunk 2 above)
Comment (bestouff) [reply src/tree.rs:26]: Yes it is, but my point is that the user will have to implement it anyway, so it probably should be reused; so having a `style()` method to ask the user how it resolved a particular `Node`'s properties looks good to me.

**Comment by bestouff** [comment]
> [...] I know that Morphorm is being looked at as a candidate [...]

You made me have a look at it. It's really cool (even if the API is a bit clunky at times) but in my small test it doesn't implement wrapping; wouldn't that be required for Taffy to use it ?

**Comment by bestouff** [comment]
Note for later: `TaffyECS` should probably be conditionally compiled, so somebody that only needs the `LayoutTree` trait could get rid of unneeded dependencies (e.g. `slotmap`). But this doesn't need to be in that PR.

File: src/tree.rs (hunk 4 above)
Comment (bestouff) [reply src/tree.rs:14]: Wouldn't `IntoIterator` consume the collection (`into_iter()` takes `self` by value) ? I think we just want using `iter()` here.

**Comment by nicoburns** [comment]
> > [...] I know that Morphorm is being looked at as a candidate [...]
> 
> You made me have a look at it. It's really cool (even if the API is a bit clunky at times) but in my small test it doesn't implement wrapping; wouldn't that be required for Taffy to use it ?

I also looked into this recently, and it does seem cool. I think Taffy would use it as another alternative algorithm (e.g. a `Display::Morphorm` mode) rather than an implementation detail for the existing one. So lacking features wouldn't preclude it. It's earlier, but I think `taffy` aims to be somewhat of a "one stop shop" for layout.

> Note for later: TaffyECS should probably be conditionally compiled, so somebody that only needs the LayoutTree trait could get rid of unneeded dependencies (e.g. slotmap). But this doesn't need to be in that PR.

This makes sense.

**Comment by nicoburns** [comment]
@jkelleyrtp What's the status of this PR? I've been basing my work off of it, and I'm concerned that it's now drifting further and further from `main` over time. My own take is that the changes in this PR are broadly good, and that while we might want to further tweak the LayoutTree trait abstraction, it might be best to get this merged and iterate on it later.

(side note: I believe that the recent force-push to `main` has broken existing pull requests and that they will need to be rebased on top of the latest main).

**Comment by alice-i-cecile** [comment]
> My own take is that the changes in this PR are broadly good, and that while we might want to further tweak the LayoutTree trait abstraction, it might be best to get this merged and iterate on it later.

This is my position as well :) I'd like to avoid long-lived branches, especially for foundational data structures.

**Comment by geom3trik** [comment]
> You made me have a look at it. It's really cool (even if the API is a bit clunky at times) but in my small test it doesn't implement wrapping; wouldn't that be required for Taffy to use it ?

Somehow missed this before. You're right that Morphorm doesn't currently support wrapping but it's something I'm planning to add for a rewrite that I'm working on at the moment (there's already a working but not very well tested implementation).

Once I've finished this rewrite then I think Morphorm will be in a better position to be integrated with taffy in some way if that's still desired.

**Comment by alice-i-cecile** [comment]
> Once I've finished this rewrite then I think Morphorm will be in a better position to be integrated with taffy in some way if that's still desired.

It is!

**Pull request closed without merging**